main/src/ca/uwaterloo/flix/util/collection/Chain.scala

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright 2023 Matthew Lutze
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ca.uwaterloo.flix.util.collection
+
+/**
+  * A linear data structure that allows fast concatenation.
+  */
+sealed trait Chain[+A] extends Iterable[A] {
+
+  /**
+    * Returns an iterator over the chain, from left to right.
+    */
+  override def iterator: Iterator[A] = this match {
+    case Chain.Empty => Iterator.empty
+    case Chain.One(x) => Iterator.single(x)
+    case Chain.Many(l, r) => Iterator.concat(l, r)
+  }
+
+  /**
+    * Concatenates `this` chain and `that` chain.
+    */
+  def ++[B >: A](that: Chain[B]) : Chain[B] = (this, that) match {
```

Comment:
This is actually slow for such a vital operation. Better to match on this and then go from there.